### ISSUE TITLE
Kubeflow Trainer SDK should not fail when a TrainJob uses a custom Kubernetes resource

### ISSUE THREAD

**Issue by LedWork** [kubeflow/sdk#516]
### What happened?

[`get_container_devices()`](https://github.com/kubeflow/sdk/blob/852519fbbbc4532ee0c9459ffd6d50daa7a5a0b9/kubeflow/trainer/backends/kubernetes/utils.py#L30) infers `device` and `device_count` from `container.resources.limits` in specific order: `gpu` -> `tpu` -> `gpu mig` -> `npu` -> `cpu` -> raise Exception.

Both below settings are valid and TrainJob is properly created for both*:

- Case 1: 
```python
trainer = CustomTrainer(
        func=...
        resources_per_node={
            "cpu": 1,
            "rdma/hca_shared_devices_a": 1,
        }
}
```

- Case 2:
```python
trainer = CustomTrainer(
        func=...
        resources_per_node={
            "rdma/hca_shared_devices_a": 1,
        }
}
```

However, for **case 2**, when the SDK reads the job back, it fails with:

```text
Unknown device type in the container resources: {'rdma/hca_shared_devices_a': IoK8sApimachineryPkgApiResourceQuantity(oneof_schema_1_validator='1', oneof_schema_2_validator=None, actual_instance='1', one_of_schemas={'str', 'float'})}
```

That means that below calls fail even if the TrainJob can be completed successfully:

- `get_job()`
- `list_jobs()`
- `wait_for_job_status()`
- `get_job_logs()`

[*source](https://github.com/Mellanox/k8s-rdma-shared-dev-plugin)

---

There is also a related UX wrinkle with mixed resources. For example:

```python
resources_per_node={
    "cpu": 4,
    "gpu": 1,
}
```

The current priority order reports this as `gpu`, while it does not need to be the actual user's intention.

### What did you expect to happen?

Unknown resources should not break SDK job/status/log access.

The SDK could just return `None` or `unknown`.

### Environment

Kubernetes version:
```bash
$ kubectl version

Client Version: v1.35.2
Kustomize Version: v5.7.1
Server Version: v1.36.0
```
Kubeflow Trainer version:
```bash
$ kubectl get pods -n kubeflow -l app.kubernetes.io/name=trainer -o jsonpath="{.items[*].spec.containers[*].image}"

ghcr.io/kubeflow/trainer/trainer-controller-manager:v2.1.0
```
Kubeflow Python SDK version:
```bash
$ pip show kubeflow

Name: kubeflow
Version: 0.4.0
```

### Impacted by this bug?

Give it a 👍 We prioritize the issues with most 👍

**Comment by tariq-hasan** [comment]
Thanks for raising the issue @LedWork.

I believe using `kubeflow.common.constants.UNKNOWN` would be the correct immediate fix for now. Adding the `trainer.kubeflow.org/device` label to the runtime and propagating it to the SDK would be the cleanest fix but that would be a separate cross-repo effort. It would be great to hear from the maintainers on the correct approach here.

/area trainer

**Comment by andreyvelich** [comment]
I agree, we should not fail if resource is unknown.
Instead of adding them as Unknown resource type, maybe we should just use the label directly, e.g. `rdma/hca_shared_devices_a`

WDYT @kubeflow/kubeflow-sdk-team @LedWork @tariq-hasan ?

**Comment by adiprathapa** [comment]
Using the resource key directly sounds right to me. It keeps the actual resource visible to the user instead of collapsing it to a generic unknown value. For the mixed case I would keep the current priority order, so a known accelerator still wins when present, and only fall back to the raw key when nothing matches. I am happy to open a PR with the fix and unit tests covering both cases from the issue if that works.

**Comment by reckless-sherixx** [comment]
@andreyvelich can I work on this issue?
